Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Plan.hs

```diff
@@ -0,0 +1,303 @@
+    paymentMode :: PaymentMode
+  }
+  deriving (Generic, ToJSON, ToSchema)
+
+data PlanFareBreakup = PlanFareBreakup
+  { component :: Text,
+    amount :: HighPrecMoney
+  }
+  deriving (Generic, ToJSON, ToSchema)
+
+data OfferEntity = OfferEntity
+  { title :: Maybe Text,
+    description :: Maybe Text,
+    tnc :: Maybe Text
+  }
+  deriving (Generic, ToJSON, ToSchema)
+
+newtype CurrentPlanRes = CurrentPlanRes
+  { currentPlanDetails :: PlanEntity
+  }
+  deriving (Generic, ToJSON, ToSchema)
+
+data PlanSubscribeRes = PlanSubscribeRes
+  { orderId :: Id DOrder.PaymentOrder,
+    orderResp :: Payment.CreateOrderResp
+  }
+  deriving (Generic, ToJSON, ToSchema)
+
+---------------------------------------------------------------------------------------------------------
+--------------------------------------------- Controllers -----------------------------------------------
+---------------------------------------------------------------------------------------------------------
+
+-- This API is for listing all the AUTO PAY plans
+planList :: (Id SP.Person, Id DM.Merchant) -> Maybe Int -> Maybe Int -> Flow PlanListAPIRes
+planList (driverId, merchantId) _mbLimit _mbOffset = do
+  plans <- Esq.runInReplica $ QPD.findByMerchantIdAndPaymentMode merchantId AUTOPAY
+  transporterConfig <- QTC.findByMerchantId merchantId >>= fromMaybeM (TransporterConfigNotFound merchantId.getId)
+  plansList <- mapM (convertPlanToPlanEntity driverId) plans
+  return $
+    PlanListAPIRes
+      { list = plansList,
+        subscriptionStartTime = transporterConfig.subscriptionStartTime
+      }
+
+-- This API is for listing current driver plan
+currentPlan :: (Id SP.Person, Id DM.Merchant) -> Flow CurrentPlanRes
+currentPlan (driverId, _merchantId) = do
+  driverInfo <- CDI.findById (cast driverId) >>= fromMaybeM (PersonNotFound driverId.getId)
+  driverPlan <- Esq.runInReplica $ QDPlan.findByDriverId driverId >>= fromMaybeM (NoCurrentPlanForDriver driverId.getId)
+  plan <- Esq.runInReplica $ QPD.findByIdAndPaymentMode driverPlan.planId (getDriverPaymentMode driverInfo.autoPayStatus) >>= fromMaybeM (PlanNotFound driverPlan.planId.getId)
+  currentPlanEntity <- convertPlanToPlanEntity driverId plan
+  return CurrentPlanRes {currentPlanDetails = currentPlanEntity}
+  where
+    getDriverPaymentMode = \case
+      Just DI.ACTIVE -> AUTOPAY
+      Just DI.SUSPENDED -> MANUAL
+      Just DI.PAUSED_PSP -> MANUAL
+      Just DI.CANCELLED_PSP -> MANUAL
+      _ -> MANUAL
+
+-- This API is to create a mandate order if the driver has not subscribed to Mandate even once or has Cancelled Mandate from PSP App.
+planSubscribe :: Id Plan -> (Id SP.Person, Id DM.Merchant) -> Flow PlanSubscribeRes
+planSubscribe planId (driverId, merchantId) = do
+  driverInfo <- CDI.findById (cast driverId) >>= fromMaybeM (PersonNotFound driverId.getId)
+  unless (isNothing driverInfo.autoPayStatus || driverInfo.autoPayStatus == Just DI.CANCELLED_PSP) $ throwError InvalidAutoPayStatus
```

Comment:
How will we know if driver cancels through PSP? Shouldn't we fetch the mandate status once before this line?